miner/worker_builder.go

```diff
@@ -0,0 +1,472 @@
+		mergedBundle.BundleGasUsed += simulatedBundle.BundleGasUsed
+	}
+
+	if len(includedTxs) == 0 {
+		return nil, nil, errors.New("include no txs when merge bundles")
+	}
+
+	mergedBundle.BundleGasPrice.Div(mergedBundle.BundleGasFees, new(big.Int).SetUint64(mergedBundle.BundleGasUsed))
+
+	return includedTxs, &mergedBundle, nil
+}
+
+// simulateBundle computes the gas price for a whole simulateBundle based on the same ctx
+// named computeBundleGas in flashbots
+func (w *worker) simulateBundle(
+	env *environment, bundle *types.Bundle, state *state.StateDB, gasPool *core.GasPool,
+	pendingTxs map[common.Address][]*txpool.LazyTransaction, currentTxCount int,
+	prune, pruneGasExceed bool,
+) (*types.SimulatedBundle, error) {
+	var (
+		tempGasUsed     uint64
+		bundleGasUsed   uint64
+		bundleGasFees   = new(big.Int)
+		ethSentToSystem = new(big.Int)
+	)
+
+	for i, tx := range bundle.Txs {
+		state.SetTxContext(tx.Hash(), i+currentTxCount)
+		sysBalanceBefore := state.GetBalance(consensus.SystemAddress)
+
+		receipt, err := core.ApplyTransaction(w.chainConfig, w.chain, &w.coinbase, gasPool, state, env.header, tx,
+			&tempGasUsed, *w.chain.GetVMConfig())
+		if err != nil {
+			log.Warn("fail to simulate bundle", "hash", bundle.Hash().String(), "err", err)
+
+			if prune {
+				if errors.Is(err, core.ErrGasLimitReached) && !pruneGasExceed {
+					log.Warn("bundle gas limit exceed", "hash", bundle.Hash().String())
+				} else {
+					log.Warn("prune bundle", "hash", bundle.Hash().String(), "err", err)
+					w.eth.TxPool().PruneBundle(bundle.Hash())
+				}
+			}
+
+			return nil, err
+		}
+
+		if receipt.Status == types.ReceiptStatusFailed && !containsHash(bundle.RevertingTxHashes, receipt.TxHash) {
+			err = errNonRevertingTxInBundleFailed
+			log.Warn("fail to simulate bundle", "hash", bundle.Hash().String(), "err", err)
+
+			if prune {
+				w.eth.TxPool().PruneBundle(bundle.Hash())
+				log.Warn("prune bundle", "hash", bundle.Hash().String())
+			}
+
+			return nil, err
+		}
+
+		bundleGasUsed += receipt.GasUsed
+
+		txGasUsed := new(big.Int).SetUint64(receipt.GasUsed)
+		txGasFees := new(big.Int).Mul(txGasUsed, tx.GasPrice())
+		sysBalanceAfter := state.GetBalance(consensus.SystemAddress)
+		sysDelta := new(big.Int).Sub(sysBalanceAfter, sysBalanceBefore)
+		sysDelta.Sub(sysDelta, txGasFees)
+		ethSentToSystem.Add(ethSentToSystem, sysDelta)
+
+		var txInPendingPool bool
+		if pendingTxs != nil {
+			sender, err := types.Sender(env.signer, tx)
+			if err != nil {
+				log.Warn("fail to simulate bundle", "hash", bundle.Hash().String(), "err", err)
+				return nil, err
+			}
+
+			if accountTxs, ok := pendingTxs[sender]; ok {
+				// check if tx is in pending pool
+				txNonce := tx.Nonce()
+
+				for _, accountTx := range accountTxs {
+					if accountTx.Tx.Tx.Nonce() == txNonce {
+						txInPendingPool = true
+						break
+					}
+				}
+			}
+		}
+
+		// If tx is not in pending pool, count the gas fees
+		if !txInPendingPool {
```

Comment:
why only count fee when `!txInPendingPool`?

Reply:
In my understanding, if a tx in this bundle is already in the pending pool, this tx will failed and the gas will not be collected right？

Reply:
no..

Reply:
Removed